xwiki-platform-core/xwiki-platform-url/xwiki-platform-url-default/src/main/java/org/xwiki/url/internal/XWikiServerClassListener.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.url.internal;
+
+import java.util.Arrays;
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.model.reference.LocalDocumentReference;
+import org.xwiki.observation.AbstractEventListener;
+import org.xwiki.observation.event.Event;
+import org.xwiki.url.URLSecurityManager;
+
+import com.xpn.xwiki.internal.event.XObjectAddedEvent;
+import com.xpn.xwiki.internal.event.XObjectDeletedEvent;
+import com.xpn.xwiki.internal.event.XObjectUpdatedEvent;
+
+/**
+ * Listener for changes on XWikiServerClass xobjects to ensure the {@link URLSecurityManager} cache is invalidated
+ * in case of change on XWikiServerClass objects.
+ *
+ * @version $Id$
+ * @since 13.3RC1
+ * @since 12.10.7
+ */
+@Component
+@Singleton
+@Named(XWikiServerClassListener.NAME)
+public class XWikiServerClassListener extends AbstractEventListener
+{
+    /**
+     * Name of the listener.
+     */
+    public static final String NAME = "org.xwiki.url.internal.XWikiServerClassListener";
+
+    private static final LocalDocumentReference XWIKISERVER_CLASS =
+        new LocalDocumentReference("XWiki", "XWikiServerClass");
```

Comment:
Small detail Ideally this would need to come from elsewhere so that if we change the class location or name we need to update less places.

Reply:
Maybe the best is to introduce WikiDescriptor/Added/Deleted/UpdatedEvent events since that what we are actually checking here (and it's not the only place where we could need that).

Reply:
> Maybe the best is to introduce WikiDescriptor/Added/Deleted/UpdatedEvent events since that's what we are actually checking here (and it's not the only place where we could need that).

Could be interesting indeed.

Reply:
I created https://jira.xwiki.org/browse/XWIKI-18552 for it.

Reply:
Note that in the meantime you could use XWikiServerClassDocumentInitializer#SERVER_CLASS unless you don't have a dependency on old core, did not checked.

Reply:
Indeed I missed it.

Reply:
Actually to use it I'd need to add a dependency to `xwiki-platform-wiki-default` and I don't really want this dependency just for that. This would be probably better if this constant would be a `xwiki-platform-wiki-api`.

Reply:
> I don't really want this dependency just for that

Definitely, I just though it was in oldcore for some reason.

> This would be probably better if this constant would be a xwiki-platform-wiki-api.

I don't know about that, I prefer not making this public too much and instead provide the missing APIs (like the events in your case which would pass the new WikiDescriptor).